### DISCUSSION TITLE
Create a different pos variable for each grid instance used in a simulation

### DISCUSSION THREAD

**Discussion by ameligrana** [mesa/mesa#1513] [Ideas]
As of now the pos variable of an agent can become buggy when in a simulation more than one grid instance is used, e.g. an agent belonging to two different networks or an agent belonging to a grid and a network. 

We could solve this by:

step 1: Identify each instance of a class in space.py with a name variable provided by the user, making it clear it should be uniquely defined between all instances of the space classes created. So we need to add a name parameter to each class instance;

step 2: assign when the agent is placed on the grid a position like agent.{id_instance_grid}+_pos.

This is a (big) breaking change, but I think a good one. This could be added to #1511.

**Comment by Corvince** [comment]
Yes, agents belonging to multiple spaces is poorly supported at the moment. 

@wang-boyu has implemented layers for Mesa-geo, maybe there are some learnings from that?

Both solutions from @Tortar seem a bit too verbose to me for simple models with only one space, but I am not sure if there is an easy way to prevent that.

**Reply by ameligrana** [reply]
Actually it is only one solution, the two points are steps, I will edit to make it clearer

**Reply by wang-boyu** [reply]
In mesa-geo there's only one space (GeoSpace) but with many layers, which is slightly different from multiple spaces that's mentioned here. In a multi-layer space, the agent's `pos` or `geometry` is the same across these layers.

Perhaps what's related to the discussion here is the `Cell` class in mesa-geo? It has a `pos` attribute for its (x, y) location. But it also lives in a `RasterLayer` which is essentially a geo-version of grid. To this end, `Cell` class also has an `indices` attribute for its (row, col) location on the grid: https://github.com/projectmesa/mesa-geo/blob/e6be96d69ff96574cb16b6694bcbf0d1f5c390b8/mesa_geo/raster_layers.py#L165-L188

I know this could be a bit confusing because in Mesa, `pos` is used for the (row, col) location on the grid, which is denoted as (x, y) in the source code. I chose this approach in mesa-geo to make `Cell` and `GeoAgent` consistent.

**Comment by ameligrana** [comment]
Another solution can be to have a dict inside each instance grid where you have the id of the agent as key and as value the position in that instance. This is simple and effective I think.

**Reply by ameligrana** [reply]
what do you think about this way instead @Corvince ?

**Reply by Corvince** [reply]
That should work, although that of course will imply some overhead for doing that. And then we have a list of agents on the grid, a list of agents on the scheduler and possibly even more places. I am wondering if there is a better way, but i have no idea yet (that's why I didn't answer before).

I think we would need to sketch out a concrete workflow for this feature to be useful.

**Reply by ameligrana** [reply]
ops, my bad my previous comments were out of touch
